### PR TITLE
[Arm64/Unix] Fix SOFTWARE_WRITE_WATCH for 64K pages

### PR DESCRIPTION
**Pull request by sdmaclea**

### CODE REVIEW

File: src/gc/softwarewritewatch.cpp [hunk 1]
```diff
@@ -10,8 +10,6 @@
 #ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
 #ifndef DACCESS_COMPILE
 
-static_assert((static_cast<size_t>(1) << SOFTWARE_WRITE_WATCH_AddressToTableByteIndexShift) == OS_PAGE_SIZE, "Unexpected OS_PAGE_SIZE");
-
```
Comment (sdmaclea) [inline src/gc/softwarewritewatch.cpp]: This change is in the 64K enable PR.

File: src/gc/softwarewritewatch.h [hunk 2]
```diff
@@ -11,6 +11,11 @@
 #ifdef FEATURE_USE_SOFTWARE_WRITE_WATCH_FOR_GC_HEAP
 #ifndef DACCESS_COMPILE
 
+#ifndef GC_PAGE_SIZE
+#define GC_PAGE_SIZE 0x1000
+#endif
```
Comment (sdmaclea) [inline src/gc/softwarewritewatch.h]: This is defined in gcpriv.h in the 64K PR, but I could not get gcpriv.h included in this file.

Comment (Maoni0) [inline src/gc/softwarewritewatch.h]: this should not be GC_PAGE_SIZE - it has no relationship with GC_PAGE_SIZE. it should be its own thing. you can name it WRITE_WATCH_UNIT_SIZE.

Comment (sdmaclea) [inline src/gc/softwarewritewatch.h]: @Maoni0 @swgillespie Done

**Comment by sdmaclea** [comment]
This will not affect [Arm64/Unix] until #10981 & #11375 is merged.  

It may be merged in any order, but will merge conflict with #10981.  

I recommend this one merge before #10981 as #10981 will not merge until after 2.0.

**Comment by sdmaclea** [comment]
@dotnet/arm64-contrib Ping can this be merged since we have passed ZBB

**Comment by BruceForstall** [comment]
@dotnet-bot test this please

**Comment by BruceForstall** [comment]
@Maoni0 @swgillespie @jkotas PTAL

**Comment by BruceForstall** [comment]
@dotnet-bot test Ubuntu arm Cross Release Build

**Comment by Maoni0** [comment]
Please also change the OS_PAGE_SIZE usage in gc_heap::revisit_written_page/gc_heap::revisit_written_pages to WRITE_WATCH_UNIT_SIZE.

**Comment by sdmaclea** [comment]
@Maoni0 Done.

**Comment by sdmaclea** [comment]
Please merge.

**Comment by briansull** [comment]
@Maoni0 PTAL

**Comment by Maoni0** [comment]
>     last_page = align_lower_page (o);

oops, my apologies @sdmaclea, this line will also need to be changed as it's still using OS_PAGE_SIZE for the alignment. would you mind creating another function called align_write_watch_lower_page that aligned lower on the WRITE_WATCH_UNIT_SIZE? 

---
Refers to: src/gc/gc.cpp:26322 in 23f7793. [](commit_id = 23f7793a7ac11a6f776aa46229c4d403f3f36d8d, deletion_comment = False)

**Comment by sdmaclea** [comment]
@Maoni I pushed your requested change.

**Comment by sdmaclea** [comment]
OS X failure is a widespread infrastructure issue. 

I think this can be merged.

**Comment by Maoni0** [comment]
Thanks for doing this work!

**Pull request merged**